### ISSUE TITLE
[DOX] Remove "using namespace seqan3" in all snippets and make all snippets Copy'n'Pastable

### ISSUE THREAD

**Issue by joshuak94** [seqan/seqan3#837]
I think it'd be more helpful to new users if the tutorial were more consistent about using the seqan3 namespace. For example, the very first code snippet in the tutorial is:
```cpp
#include <seqan3/io/stream/debug_stream.hpp> // for debug_stream
using namespace seqan3;
int main()
{
    debug_stream << "Hello World!\n";
}
```
Then, in code snippet a bit below it, ```seqan3::argument_parser``` is used. We should be consistent with prefacing everything with ```seqan3::``` so that it's clear which parts of the code are from seqan.

**Comment by smehringer** [comment]
I agree and I think we should never use `using namespace seqan3;`

**Comment by rrahn** [comment]
> I agree and I think we should never use using namespace seqan3;

I think we should always use it. 
Including the namespace is easy to do and makes many code examples so much easier to read. 
If people start with SeqAn and learn about it they either just start with SeqAn (no other library except std) or they already worked with c++ and other libraries and are aware of the `using namespace` implications.
But we need to make sure, that all stand alone snippets then also have a using namespace seqan3 declaration inside.

```cpp
int main()
{
{
//! [snippet]
using namespace seqan3;
auto v = view::make_me_happy(...);
//! [snippet]

}
}
```

**Comment by eseiler** [comment]
I agree with @rrahn. It might be fine to use `seqan3::` for some of the tutorials, but later on it will not add to the comprehension. (configurations where all elements are prefixed with `seqan3::` are not really easy to read)

**Comment by h-2** [comment]
We should also consider whether valid links to the api-documentation are generated from our code snippets...

**Comment by h-2** [comment]
We should discuss it at the next strategy meeting

**Comment by smehringer** [comment]
We decided on the following:

The API dox will *never* have using namespace.
The beginner tutorials are allowed to have `using namespace seqan3;` for simplicity but we should add a note box in the beginnning that this is not recommended for app development etc but only done for simplicity.
The How to will also never have using namespace.

**Comment by joshuak94** [comment]
Cool! When you say
> The API dox will _never_ have using namespace.
Does this include snippets?

**Comment by smehringer** [comment]
> Does this include snippets?

Yes

**Comment by joshuak94** [comment]
Sounds good! Sounds like a pull request should be made fixing all of the snippets then... Should I get started on that?

**Comment by h-2** [comment]
> Note: When using our literals you will need to include them explicitly with
> `using seqan3::operator""_dna4;`

Wow, this is horrible, I never thought about that :open_mouth:

**Comment by smehringer** [comment]
> Should I get started on that?

That would be great :) 
Best make small PR's (maybe per module) so they don't take too long to be merged.

**Comment by h-2** [comment]
If you do start working on this, also make the snippets full text, i.e. include main function and includes so the snippet is fully copy'n'pastable.

**Comment by joshuak94** [comment]
Wait. If we want to make the snippets copy & pastable, how does this work for snippet files which include multiple snippets inside them? Example:

```cpp

#include <blah>

int main()
{
{
//! [snippet1]
do something
//! [snippet 1]
}

{
//! [snippet2]
do something else
//! [snippet2]
}
}
```

**Comment by rrahn** [comment]
@joshuak94 we still need this. can you give an estimate whether or not you can work on this in June?

**Comment by joshuak94** [comment]
I can work on it!

**Comment by joshuak94** [comment]
@rrahn My question above was never answered, though. @h-2 said the whole thing should be copy'n'pastable, so the snippet should contain the include statements and the main function. What about snippet files containing two or more snippets? Should those snippets all get their own file, such that there is always one file per snippet code?

**Comment by joshuak94** [comment]
@rrahn @h-2 Bumping for my last question regarding multiple snippets in the same file..

**Comment by h-2** [comment]
Yes, the snippets should be adapted so that they are individually valid. This can mean making more individual files, but it can also mean merging some snippets into one larger snippet. This depends on the context and is mostly up to you I would say!

**Comment by joshuak94** [comment]
@h-2 Sounds good! A few more questions.

1. If the snippets are now going to be one snippet per file, do we need to use `[[snippet]]` notation? Or should I just get rid of those and do `\include` instead of `\snippet`.

2. Some of the examples use things like try catch blocks or `(void)` for unused variables. I'm guessing it's not a problem if those are visible in the examples? And would `(void)` be better, or `[[maybe_unused]]`.

**Comment by eseiler** [comment]
> @h-2 Sounds good! A few more questions.
> 
> 1. If the snippets are now going to be one snippet per file, do we need to use `[[snippet]]` notation? Or should I just get rid of those and do `\include` instead of `\snippet`.

If you want to include the whole file, `\snippet` and `\include` are equivalent. So I would use `\include` instead of marking the whole file as a snippet.

> 2. Some of the examples use things like try catch blocks or `(void)` for unused variables. I'm guessing it's not a problem if those are visible in the examples? And would `(void)` be better, or `[[maybe_unused]]`.

You should be able to remove `(void)` and `[[maybe_unused]]`. We added `-Wno-unused-variable` to the compile flags of the snippet tests.
If you somewhere get unused parameter warnings, we can discuss if we want to add `-Wno-unused-parameter` as well.

**Comment by joshuak94** [comment]
Ah, great. Thanks!

**Comment by joshuak94** [comment]
Should anything be changed with snippets that do not work, and are in comment blocks? Or is it enough to have them be as follows:
```cpp
int main()
{
// does not work:
// seqan3::dna4 my_letter{0};      // we want to set the default, an A
// seqan3::dna4 my_letter{'A'};    // we also want to set an A, but we are setting value 65

// std::cout << my_letter; // you expect 'A', but how would you access the number?
}
```

**Comment by eseiler** [comment]
IIRC we wanted to remove the "does not work" stuff.

**Comment by joshuak94** [comment]
>You should be able to remove (void) and [[maybe_unused]]. We added -Wno-unused-variable to the compile flags of the snippet tests.
If you somewhere get unused parameter warnings, we can discuss if we want to add -Wno-unused-parameter as well.

What about -Wno-unused-but-set-variable? For something like this:
```cpp
    bool is_unpaired_char = '.'_wuss51.is_unpaired();       // true
    is_unpaired_char = seqan3::is_unpaired('{'_wuss51);     // false
```

**Comment by joshuak94** [comment]
Also, is it just snippets in the API documentation that should be Copy'n'Pastable? Or is it everything. For example, [this](http://docs.seqan.de/seqan/3-master-dev/classseqan3_1_1arithmetic__range__validator.html) snippet is fine, but in the tutorial where it mentions the `arithmetic_range_validator`, there's a small snippet which just shows how to construct the validator (http://docs.seqan.de/seqan/3-master-dev/tutorial_argument_parser.html). Should that also be copy'n'pastable?

**Comment by h-2** [comment]
I think we should add `-Wno-unused-parameter` and `-Wno-unused-but-set-variable`

> but in the tutorial where it mentions the

The tutorials shall keep `using namespace seqan3;`.

**Comment by joshuak94** [comment]
> The tutorials shall keep `using namespace seqan3;`.

So tutorials can be excerpts of code, and not fully copy'n'pastable?

**Comment by h-2** [comment]
> So tutorials can be excerpts of code, and not fully copy'n'pastable?

Hm, that's a different question :thinking: But IIRC correctly that's what we decided on.

Since the tutorials are always done en-bloc and are meant to be followed from beginning to end, it's ok to have smaller snippets that constitute a whole and it reducses visual noise to omit `seqan3::`. The API docs on the other hand are meant as a resource to quickly lookup the properties of an interface and scanning over the the documentation to find what you need. Therefore the snippets should be self-contained and most-easily copy'n'pastable into user code.

**Comment by joshuak94** [comment]
Hmmmm okay. But are we *requiring* then, that tutorial snippets omit `seqan3::`? Because some of the tutorial snippets are excerpts from larger examples (see snippet/argument_parser/validators_1.cpp), and if the entire example requires `seqan3::`, I can't really omit it for just the tutorial part without creating a redundant file.

**Comment by joshuak94** [comment]
@h-2 Also `-Werror=unused-local-typedefs`?

**Comment by h-2** [comment]
Yep. All the unused stuff can be ignored for snippets.

**Comment by joshuak94** [comment]
@h-2 Do you know if every gcc flag has a "no" option? i.e. if I just did "-Wno-unused", would that be sufficient?

**Comment by smehringer** [comment]
Fixed by #1230 #1198 #1197 #1147 #1132 #1127 #1124 #1115
